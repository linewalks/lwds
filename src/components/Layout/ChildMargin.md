#### Horizontal

```js
import Button from 'components/Button/Button'
const Horizontal = () => {
  return (
    <>
      <h5>Margin 8 (default)</h5>
      <ChildMargin>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
      </ChildMargin>
      <h5>Margin 16 (default)</h5>
      <ChildMargin margin={16}>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
      </ChildMargin>
    </>
  )
}
Horizontal()
```

#### Vertical

```js
import Button from 'components/Button/Button'
const Vertical = () => {
  return (
    <>
      <h5>Margin 8 (default)</h5>
      <ChildMargin orientation="vertical">
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
      </ChildMargin>
      <h5>Margin 16 (default)</h5>
      <ChildMargin orientation="vertical" margin={16}>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
        <Button variant="primary">primary</Button>
      </ChildMargin>
    </>
  )
}
Vertical()
```
