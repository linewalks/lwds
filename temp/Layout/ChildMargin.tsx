import React from 'react'
import styled from 'styled-components'

const WrapLayout = styled.div`
  display: flex;
  ${({ orientation, margin }: { orientation: string; margin: number }): any => {
    return orientation === 'vertical'
      ? `flex-direction: column;
      & > * {
        margin-bottom: ${margin}px;
      }
      & > *:last-child {
        margin-bottom: 0;
      }`
      : `flex-direction: row;
      & > * {
        margin-right: ${margin}px;
      }
      & > *:last-child {
        margin-right: 0;
      }`
  }}
`

interface ChildMarginProps {
  className: string
  orientation: string | 'horizontal' | 'vertical'
  margin: number
  style?: object
  children: React.ReactNode
}

const ChildMargin = (props: ChildMarginProps): any => {
  const { className, orientation, margin, style, children, ...rest } = props

  return (
    <WrapLayout
      className={className}
      orientation={orientation}
      margin={margin}
      style={style}
      {...rest}
    >
      {children}
    </WrapLayout>
  )
}

ChildMargin.defaultProps = {
  orientation: 'horizontal',
  margin: 8,
}

export default ChildMargin
