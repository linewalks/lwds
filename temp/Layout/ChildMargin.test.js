import React from 'react'
import each from 'jest-each'
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'
import 'jest-styled-components'

import ChildMargin from '@components/Layout/ChildMargin'

describe('vertical', () => {
  each([
    4,
    8,
    16
  ]).it('%d margin',  (margin) => {
    const tree = renderer.create(
      <ChildMargin orientation="vertical" margin={margin}>
        <div>child</div>
        <div>child</div>
        <div>child</div>
      </ChildMargin>
    ).toJSON()

    expect(tree).toHaveStyleRule('flex-direction', 'column')
    expect(tree).toMatchSnapshot()
  })
})

describe('horizontal', () => {
  each([
    4,
    8,
    16
  ]).it('%d margin',  (margin) => {
    const tree = renderer.create(
      <ChildMargin orientation="horizontal" margin={margin}>
        <div>child</div>
        <div>child</div>
        <div>child</div>
      </ChildMargin>
    ).toJSON()

    expect(tree).toHaveStyleRule('flex-direction', 'row')
    expect(tree).toMatchSnapshot()
  })
})
