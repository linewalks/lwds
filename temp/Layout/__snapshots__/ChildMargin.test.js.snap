// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`horizontal 4 margin 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 > * {
  margin-right: 4px;
}

.c0 > *:last-child {
  margin-right: 0;
}

<div
  className="c0"
  orientation="horizontal"
>
  <div>
    child
  </div>
  <div>
    child
  </div>
  <div>
    child
  </div>
</div>
`;

exports[`horizontal 8 margin 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 > * {
  margin-right: 8px;
}

.c0 > *:last-child {
  margin-right: 0;
}

<div
  className="c0"
  orientation="horizontal"
>
  <div>
    child
  </div>
  <div>
    child
  </div>
  <div>
    child
  </div>
</div>
`;

exports[`horizontal 16 margin 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 > * {
  margin-right: 16px;
}

.c0 > *:last-child {
  margin-right: 0;
}

<div
  className="c0"
  orientation="horizontal"
>
  <div>
    child
  </div>
  <div>
    child
  </div>
  <div>
    child
  </div>
</div>
`;

exports[`vertical 4 margin 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > * {
  margin-bottom: 4px;
}

.c0 > *:last-child {
  margin-bottom: 0;
}

<div
  className="c0"
  orientation="vertical"
>
  <div>
    child
  </div>
  <div>
    child
  </div>
  <div>
    child
  </div>
</div>
`;

exports[`vertical 8 margin 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > * {
  margin-bottom: 8px;
}

.c0 > *:last-child {
  margin-bottom: 0;
}

<div
  className="c0"
  orientation="vertical"
>
  <div>
    child
  </div>
  <div>
    child
  </div>
  <div>
    child
  </div>
</div>
`;

exports[`vertical 16 margin 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 > * {
  margin-bottom: 16px;
}

.c0 > *:last-child {
  margin-bottom: 0;
}

<div
  className="c0"
  orientation="vertical"
>
  <div>
    child
  </div>
  <div>
    child
  </div>
  <div>
    child
  </div>
</div>
`;
